core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -195,34 +212,69 @@ public void beforeListVariableChanged(ListVariableDescriptor<Solution_> variable
 
     public void afterListVariableChanged(ListVariableDescriptor<Solution_> variableDescriptor, Object entity, int fromIndex,
             int toIndex) {
-        Collection<ListVariableListenerNotifiable<Solution_>> notifiables = notifiableRegistry.get(variableDescriptor);
+        var notifiables = notifiableRegistry.get(variableDescriptor);
         if (!notifiables.isEmpty()) {
             ListVariableNotification<Solution_> notification = Notification.listVariableChanged(entity, fromIndex, toIndex);
-            for (ListVariableListenerNotifiable<Solution_> notifiable : notifiables) {
+            for (var notifiable : notifiables) {
                 notifiable.notifyAfter(notification);
             }
             notificationQueuesAreEmpty = false;
         }
+        listVariableEventList.add(new ListVariableEvent(entity, fromIndex, toIndex));
     }
 
     public void triggerVariableListenersInNotificationQueues() {
-        for (Notifiable notifiable : notifiableRegistry.getAll()) {
+        for (var notifiable : notifiableRegistry.getAll()) {
             notifiable.triggerAllNotifications();
         }
+        if (listVariableDescriptor != null && !cascadingUpdateShadowVarDescriptorList.isEmpty()) {
+            triggerCascadingUpdateShadowVariableUpdate();
+        }
         notificationQueuesAreEmpty = true;
+        listVariableEventList.clear();
+    }
+
+    /**
+     * Triggers all cascading update shadow variable user-logic.
+     */
+    private void triggerCascadingUpdateShadowVariableUpdate() {
+        if (listVariableEventList.isEmpty() || cascadingUpdateShadowVarDescriptorList.isEmpty()) {
+            return;
+        }
+        for (var cascadingUpdateShadowVariableDescriptor : cascadingUpdateShadowVarDescriptorList) {
+            for (var event : listVariableEventList) {
+                var values = listVariableDescriptor.getValue(event.entity());
+                // Evaluate all elements inside the range
+                evaluateFromIndex(values, event.fromIndex(), event.toIndex(), true, cascadingUpdateShadowVariableDescriptor);
+                // Evaluate later elements, but stops when there is no change
+                evaluateFromIndex(values, event.toIndex(), values.size(), false, cascadingUpdateShadowVariableDescriptor);
```

Comment:
I find it surprising that this is faster than the alternatives discussed.
If there's two events to say `process the entire list`, this will do it twice, even though it's unnecessary.

Reply:
Right. However, the default moves do not generate two events when the source and destination entities are the same and we were executing unnecessary logic because it is not a valid state for the built-in moves.

It may be possible for the custom moves to generate two events, but it still seems incorrect to me to duplicate events as it would require additional work to create before/after events and undo the state.

Reply:
Technically `CompositeMove` _could_ run any number of moves at the same time. But you're right - this is a corner case and the mechanism needs to work well for the likely cases.